# Final Project: Simple Statistical Analysis Package for Bivariate Data

## Why choose to do statistical analysis?
I am currently also taking COGS14B, which is a course focused on statistical analysis methods, including but not limited to methods for analyzing bivariate data, such as the well-known Student's t-test, z-test, and such. I thought it would benefit my understanding of materials from both courses if I were able to break down the logic of those statistical analysis methods and build them back, line by line, into a fully-functional Python program. I will hereby present my work below:

In [1]:
# This cell gives an overview of the modules imported from the standard Python library
# and those imported as my original creation.

# This is the main statistical analysis package.
from simple_statistical_analysis_package_for_bivariate_data import student_t_tests, z_test, linear_regression, finding_correlation_coefficient_r 
# Good old friends from the standard Python library.
import string
import math
import numpy as np
import re

If you would what like to find out more in detail about what each of the statistical analysis method does, check out the docstrings for the functions by typing in an empty cell one of the functions you are interested in--`student_t_tests`, `z_test`, `linear_regression`, or `finding_correlation_coefficient_r`--then finish the line with **?**.

### Let's first check out the z-test as a starter.
Suppose you have an idea about the general class average on a midterm being 79% with a standard deviation of 3%. You then draw a convenience sample of 10 students, who are all in your study group, and you are interested in whether your group has performanced significantly different from the class on average since you guys attended the same review session and together created the same cheat sheet. What you collected at hand is a sample mean score of 86%, and luckily knowing these four data--population mean, population standard deviation, sample mean, and sample size--is enough for performing a z-test to answer your question.

In [2]:
z_test(79, 3, 86, 10)

'The calculated z-statistic is: 7.378647873726218'

### Then, let's see a few t-test scenarios.
You will be prompted to provide the data that you would like to be analyzed. If you don't have anything right off the top of your head, try type in a few random numbers, or paste in the following demo datasets sequentially for each t-test prompt.

87 84 83 86.8 89 87.5 87.25 80 91 91 84.5 80

81

105 119 100 97 96 101 94 95 98

96 99 94 89 96 93 88 105 88

9 8 7 7 8 9 7 7 8

7 7 3 8 6 4 6 7 4

In [3]:
student_t_tests()

Which type of t-test are you trying to perform:
	A. Single-sample t-test.
	B. Independent samples t-test.
	C. Dependent samples t-test.
a
Please paste your single-sample data in the form of a list below:
87 84 83 86.8 89 87.5 87.25 80 91 91 84.5 80
What is the population mean that your data is aiming to estimate?
81


'Here is your calculated t_statistic: 4.59131333857016'

If you select option A choose to copy and paste in the frist two demo datasets when propmted, you output should be **'Here is your calculated t_statistic: 4.59131333857016'**.

In [4]:
student_t_tests()

Which type of t-test are you trying to perform:
	A. Single-sample t-test.
	B. Independent samples t-test.
	C. Dependent samples t-test.
b
Please paste your first dataset to the independent samples in the form of a list below:
105 119 100 97 96 101 94 95 98
Please paste your second dataset to the independent samples in the form of a list below:
96 99 94 89 96 93 88 105 88


'Here is your calculated t_statistic: 1.9947880650265368'

If you select option B choose to copy and paste in the third and fourth demo datasets when propmted, you output should be **'Here is your calculated t_statistic: 1.9947880650265368'**.

In [5]:
student_t_tests()

Which type of t-test are you trying to perform:
	A. Single-sample t-test.
	B. Independent samples t-test.
	C. Dependent samples t-test.
c
Please paste your pre-treatment dataset in the form of a list below:
9 8 7 7 8 9 7 7 8
Please paste your post-treatment dataset in the form of a list below:
7 7 3 8 6 4 6 7 4


'Here is your calculated t_statistic: -3.0'

If you select option C choose to copy and paste in the last two demo datasets when propmted, you output should be **'Here is your calculated t_statistic: -3.0'**.

### Now, the classical linear regression and correlational analysis that would be performed on any set of bivariate data.

Suppose you want to find out if there's any association between study hours and test performances. You obtained two datasets from surveying your friends about how they did on the MATH20C midterm, one dataset contains their student hour and the other has their corresponding grades.

In [6]:
study_hours = [11, 3, 9, 5, 1, 7]
test_grades = [91, 77, 90, 80, 75, 85]

In [7]:
linear_regression(study_hours, test_grades)

The line of linear regression for these two datasets is y = 1.7714285714285714x + 72.37142857142857


In [8]:
finding_correlation_coefficient_r(study_hours, test_grades)

The Pearson's correlation coefficient for this pair of data is 0.9858672630886147, indicating a strong association between the two.


### Lastly, I will wrap up by calling the test functions designed for debugging my statistical analysis functions in ways that I can come up with to make sure that the functions are validated bugs-free.

In [9]:
!pytest test_functions.py

============================= test session starts ==============================
platform darwin -- Python 3.8.3, pytest-5.4.3, py-1.9.0, pluggy-0.13.1
rootdir: /Users/Muchan/cogs18_final_project/modules
collected 6 items                                                              

test_functions.py ......                                                 [100%]

============================== 6 passed in 0.30s ===============================


### Link to all project codes on GitHub: 
https://github.com/Lord-of-Bugs/cogs18_final_project_fa20